lib/axon/losses.ex

```diff
@@ -958,6 +958,55 @@ defmodule Axon.Losses do
     )
   end
 
+  @doc ~S"""
+  Cosine Similarity error loss function.
+
+  $$l_i = \sum_i (\hat{y_i} - y_i)^2$$
+
+  ## Argument Shapes
+
+    * `y_true` - $(d_0, d_1, ..., d_n)$
+    * `y_pred` - $(d_0, d_1, ..., d_n)$
+
+  ## Options
+
+    * `:reduction` - reduction mode. One of `:mean`, `:sum`, or `:none`.
+      Defaults to `:none`.
+    * `:axes` - Defaults to `[1]`.
+    * `:eps` - Defaults to `1.0e-6`.
+
+  ## Examples
+
+      iex> y_pred = Nx.tensor([[1.0, 0.0], [1.0, 1.0]])
+      iex> y_true = Nx.tensor([[0.0, 1.0], [1.0, 1.0]])
+      iex> Axon.Losses.cosine_similarity(y_true, y_pred)
+      #Nx.Tensor<
+        f32[2]
+        [0.0, 1.0]
+      >
+  """
+
+  defn cosine_similarity(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, axes: [1], eps: 1.0e-6, reduction: :none)
+    axes = opts[:axes]
+    eps = opts[:eps]
+
+    w12 = Nx.sum(y_true * y_pred, axes: axes)
+    w1 = Nx.LinAlg.norm(y_true, axes: axes)
+    w2 = Nx.LinAlg.norm(y_pred, axes: axes)
+    n12 = Nx.max(w1 * w2, eps)
+    loss = w12 / n12
```

Comment:
@seanmor5 WDYT about moving the basic cosine similarity block to Scholar instead of keeping it here? We could actually do this with all of the loss functions here, in that case.

It would become something like:

```elixir
loss = Scholar.Metrics.cosine_similarity(y_true, y_pred, opts)
transform(...)
```

Reply:
@polvalente I am not yet ready to have Axon completely depend on Scholar, but in the future when Scholar gets built out more I think it makes sense